Upgrade to Infinispan 14.0.4.Final

### Description

We need to upgrade to ISPN 14.0.4.Final where JBoss Marshalling will be upgrated to 2.1.1.Final. This will fix `InaccessibleObjectException` on JDK 17 https://github.com/keycloak/keycloak/issues/15607.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

> @Tristan971 Could you please provide a few more detail?

Small use of private api as per the stack trace in #15607 

> Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected java.util.AbstractMap() accessible: module java.base does not "opens java.util" to unnamed module @396a51ab

Was tracked down to a transitive dependency of Infinispan which is pulling and older jboss-marshalling-osgi version transitively.

While the original reporter experienced an LDAP-related issue, in my case this exception triggered while trying to delete a realm user. I didn't push the exploration of the symptoms further considering how many things go through the Infinispan caching layer when using clustered KC instances, as I suspect much more would have been broken. But this is merely suspicion.

--- a/model/infinispan/src/main/java/org/keycloak/cluster/infinispan/InfinispanNotificationsManager.java
+++ b/model/infinispan/src/main/java/org/keycloak/cluster/infinispan/InfinispanNotificationsManager.java
@@ -258,9 +258,15 @@ public class InfinispanNotificationsManager {
 
                 });
             } catch (RejectedExecutionException ree) {
-                // avoid touching the cache when creating a log message to avoid a deadlock in Infinispan 12.1.7.Final
-                logger.errorf("Rejected submitting of the event for key: %s. Server going to shutdown or pool exhausted. Pool: %s", key, listenersExecutor.toString());
-                throw ree;
+                // server is shutting down or pool was terminated - don't throw errors
+                if (ree.getMessage() != null && (ree.getMessage().contains("Terminated") || ree.getMessage().contains("Shutting down"))) {
+                    logger.warnf("Rejected submitting of the event for key: %s because server is shutting down or pool was terminated.", key);
+                    logger.debug(ree);
+                } else {
+                    // avoid touching the cache when creating a log message to avoid a deadlock in Infinispan 12.1.7.Final
+                    logger.errorf("Rejected submitting of the event for key: %s. Server going to shutdown or pool exhausted. Pool: %s", key, listenersExecutor.toString());
+                    throw ree;
+                }
             }
         }
 

--- a/model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java
+++ b/model/infinispan/src/main/java/org/keycloak/connections/infinispan/DefaultInfinispanConnectionProviderFactory.java
@@ -393,7 +393,7 @@ public class DefaultInfinispanConnectionProviderFactory implements InfinispanCon
 
     // Used for cross-data centers scenario. Usually integration with external JDG server, which itself handles communication between DCs.
     private void configureRemoteCacheStore(ConfigurationBuilder builder, boolean async, String cacheName) {
-        String jdgServer = config.get("remoteStoreHost", "localhost");
+        String jdgServer = config.get("remoteStoreHost", "127.0.0.1");
         Integer jdgPort = config.getInt("remoteStorePort", 11222);
 
         // After upgrade to Infinispan 12.1.7.Final it's required that both remote store and embedded cache use
@@ -422,7 +422,7 @@ public class DefaultInfinispanConnectionProviderFactory implements InfinispanCon
     }
 
     private void configureRemoteActionTokenCacheStore(ConfigurationBuilder builder, boolean async) {
-        String jdgServer = config.get("remoteStoreHost", "localhost");
+        String jdgServer = config.get("remoteStoreHost", "127.0.0.1");
         Integer jdgPort = config.getInt("remoteStorePort", 11222);
 
         // After upgrade to Infinispan 12.1.7.Final it's required that both remote store and embedded cache use

--- a/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remotestore/ClientListenerExecutorDecorator.java
+++ b/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remotestore/ClientListenerExecutorDecorator.java
@@ -127,8 +127,15 @@ public class ClientListenerExecutorDecorator<K> {
         } catch (RejectedExecutionException ree) {
             eventsInProgress.remove(key);
 
-            logger.errorf("Rejected execution of task for the event '%s' . Try to increase the pool size. Pool is '%s'", event.toString(), decorated.toString());
-            throw ree;
+            // server is shutting down or pool was terminated - don't throw errors
+            if (ree.getMessage() != null && (ree.getMessage().contains("Terminated") || ree.getMessage().contains("Shutting down"))) {
+                logger.warnf("Rejected execution of task for the event '%s' because server is shutting down or pool was terminated.", event.toString());
+                logger.debug(ree);
+            } else {
+                // avoid touching the cache when creating a log message to avoid a deadlock in Infinispan 12.1.7.Final
+                logger.errorf("Rejected execution of task for the event '%s' . Try to increase the pool size. Pool is '%s'", event.toString(), decorated.toString());
+                throw ree;
+            }
         }
     }
 

--- a/pom.xml
+++ b/pom.xml
@@ -81,8 +81,8 @@
         <jakarta.persistence.version>2.2.3</jakarta.persistence.version>
         <hibernate-orm.version>5.6.14.Final</hibernate-orm.version>
         <hibernate.c3p0.version>${hibernate-orm.version}</hibernate.c3p0.version>
-        <infinispan.version>14.0.2.Final</infinispan.version>
-        <infinispan.protostream.processor.version>4.5.0.Final</infinispan.protostream.processor.version>
+        <infinispan.version>14.0.4.Final</infinispan.version>
+        <infinispan.protostream.processor.version>4.6.0.Final</infinispan.protostream.processor.version>
         <jackson.version>2.13.4</jackson.version>
         <jackson.databind.version>2.13.4.2</jackson.databind.version>
         <jackson.annotations.version>${jackson.version}</jackson.annotations.version>

